### DISCUSSION TITLE
Reading a remote file

### DISCUSSION THREAD

**Discussion by BrunoTeixeira1996** [gokrazy/gokrazy#240] [Q&A] [answered]
Hi guys,
I need to read a file that is not stored localy inside gokrazy. Instead that file lives elsewhere in another server (on the same LAN).
So I was wondering if it is possible to read that remote file in gokrazy without having to copy to the `/perm` partition (because that file will be modified so I need to have the correct information and not past information)

**Comment by stapelberg** [comment]
In your program, you could just download the file contents via HTTP, and install a web server on the other machine? That sounds easiest to me.

I think you should describe more specifically what you’re trying to achieve, as it’s pretty hard to give good recommendations otherwise.

**Reply by BrunoTeixeira1996** [reply]
That would work but i didnt want to install anything on the remote server . The main idea is to have a gokrazy appliance that centralizes logs. So if i have a bot running on a remote web server and i want to Access the logs i could just use gokrazy to check those without having a vpn or something. 

Like i said, the web server would work (by exposing the log file and then read using http) but i would like to rsync or sftp the file from web server to gokrazy without installing anything in the remote web server.

Not sure if i was clear enough, but tell me if you want another example.

Thanks

**Reply by damdo** [reply]
There exists a [native go implementation of rsync](https://github.com/gokrazy/rsync), from @stapelberg (!), which could be leveraged here to fetch logs via rsync over SSH. Although I'm not sure whether that tool needs the `ssh` binary bindings or uses the Go SSH library to perform the operation. If the former is true, you might not be able to use it that way (gokrazy doesn't have the ssh binary).

Another option, and probably not the most straightforward thing to do, considering nowadays there is a lot of standardisation around log centralisation and processing (e.g. the OpenTelemetry), is to write something yourself.
Provided the remote server has an ssh server, you could leverage SSH, and write a pure go client that fetches logs via that and centralizes them on the gokrazy device.
A good starting point could be: https://medium.com/@marcus.murray/go-ssh-client-shell-session-c4d40daa46cd

**Reply by BrunoTeixeira1996** [reply]
Yes that seems reasonable but @stapelberg mentioned about syslog. Right now my bot writes to a file choosen by me but maybe if I write that to syslog I could make use of https://github.com/gokrazy/syslogd to read logs from a remote server to gokrazy right? I think this is the best approach here, but I am curious of what you guys think

**Reply by stapelberg** [reply, marked as answer]
If you control both sides and want an easy way to collect log lines from different systems in the same local network, I think remote syslog is the best option.

**Comment by BrunoTeixeira1996** [comment]
@stapelberg and @damdo i've built something https://github.com/BrunoTeixeira1996/log-viewer for this.
I expose journalctl and then can view all logs inside gokrazy.
Thank you for the input!
